src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
+      this._onError(error.status, error.error);
+    });
+  }
+
+  /**
+   * POST /api/kernels/[:kernel_id]/restart
+   *
+   * Restart the kernel.
+   */
+  restart(): Promise<any> {
+    this._handleStatus('restarting');
+    this._stopChannels();
+
+    var url = utils.urlJoinEncode(this._kernelUrl, 'restart');
+    return utils.ajaxRequest(url, {
+      method: "POST",
+      dataType: "json",
+      useHeader: 'Location'
+    }).then((data: any) => {
+      this._kernelCreated(data);
+      this._onSuccess(data);
+    }, (error) => {
+      this._kernelDead();
+      this._onError(error.status, error.error);
+    });
+  }
+
+  /**
+   * Reconnect to a disconnected kernel. This is not actually a
+   * standard HTTP request, but useful function nonetheless for
+   * reconnecting to the kernel if the connection is somehow lost.
+   */
+  reconnect(): void {
+    if (this.isConnected) {
+      return;
+    }
+    this._reconnectAttempt = this._reconnectAttempt + 1;
+    this._handleStatus('reconnecting');
+    this._startChannels();
+  }
+
+  /**
+   * Send a message on the Kernel's shell channel.
+   */
+  sendShellMessage(msg_type: string, content: any, metadata = {}, buffers: string[] = []): IKernelFuture {
+    if (!this.isConnected) {
+      throw new Error("kernel is not connected");
+    }
+    var msg = this._getMsg(msg_type, content, metadata, buffers);
+    msg.channel = 'shell';
+
+    this._ws.send(serialize.serialize(msg));
+
+    var future = new KernelFutureHandler(() => {
+      this._handlerMap.delete(msg.header.msgId);
+    });
+
+    this._handlerMap.set(msg.header.msgId, future);
+
+    return future;
+  }
+
+  /**
+   * Get kernel info.
+   *
+   * Returns a KernelFuture that will resolve to a `kernel_info_reply` message documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#kernel-info)
+   */
+  kernelInfo(): IKernelFuture {
+    return this.sendShellMessage("kernel_info_request", {});
+  }
+
+  /**
+   * Get info on an object.
+   *
+   * Returns a KernelFuture that will resolve to a `inspect_reply` message documented
+   * [here](http://ipython.org/ipython-doc/dev/development/messaging.html#object-information)
+   */
+  inspect(code: string, cursor_pos: number): IKernelFuture {
+    var content = {
+      code: code,
+      cursor_pos: cursor_pos,
+      detail_level: 0
+    };
+    return this.sendShellMessage("inspect_request", content);
+  }
+
+  /**
+   * Execute given code into kernel, returning a KernelFuture.
+   *
+   * @example
+   *
+   * The options object should contain the options for the execute
+   * call. Its default values are:
+   *
+   *      options = {
+   *        silent : true,
+   *        user_expressions : {},
+   *        allow_stdin : false,
+            store_history: false
+   *      }
+   *
+   */
+  execute(code: string, options?: { silent?: boolean; user_expressions?: any; allow_stdin?: boolean; store_history?: boolean; }): IKernelFuture {
```

Comment:
I think it would make sense to break that options type out into a separate interface.